src/mds/Server.cc

```diff
@@ -420,12 +420,11 @@ void Server::handle_client_session(MClientSession *m)
     
   case CEPH_SESSION_REQUEST_CLOSE:
     {
-      if (session->is_closed() || 
-	  session->is_closing() ||
-	  session->is_killing()) {
-	dout(10) << "already closed|closing|killing, dropping this req" << dendl;
-	m->put();
-	return;
+      if (session->is_closed() || session->is_closing() || session->is_killing()) {
+        dout(10) << "session alredy closed|closing|killing, responding with CEPH_SESSION_CLOSE" << dendl;
+        mds->send_message_client(new MClientSession(CEPH_SESSION_CLOSE), session);
+        m->put();
+        return;
```

Comment:
I was coming across few issues to which Patrick found this solution. You can refer the following comments -
https://github.com/ceph/ceph/pull/21065#issuecomment-380389137
https://github.com/ceph/ceph/pull/21065#issuecomment-380907455

Reply:
Here the session is already killed or killing because of the eviction, so i don't think it's properly for the mds to reply to the evicted client here. we could just let the hang request be cleaned after the client realize it is in that blacklist.

Reply:
@taodd That would be possible but I don't see a downside to the MDS telling the client to close its session.

@ukernel do you see a problem with this?

Reply:
when client requests session close, it should have no dirty caps. It should be harmless that mds blindly sends a reply.  kernel client handles this case by using timed wait. I personally prefer kernel's appocah